src/main/java/org/jenkinsci/plugins/gitclient/GitCredentialBindings.java

```diff
@@ -0,0 +1,76 @@
+package org.jenkinsci.plugins.gitclient;
+
+import com.cloudbees.plugins.credentials.common.StandardCredentials;
+import hudson.FilePath;
+
+import hudson.model.Executor;
+import hudson.model.Node;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.plugins.git.GitTool;
+import hudson.tools.ToolDescriptor;
+import hudson.tools.ToolInstallation;
+import hudson.tools.ToolLocationNodeProperty;
+import jenkins.model.Jenkins;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+import java.io.IOException;
+import java.util.List;
+
+public interface GitCredentialBindings {
+
+    void setKeyBindings(@NonNull StandardCredentials credentials);
+
+    void setRunEnviornmentVariables(@NonNull FilePath filePath, @NonNull TaskListener listener) throws IOException, InterruptedException;
+
+    GitClient getGitClientInstance(TaskListener listener) throws IOException, InterruptedException;
+
+    default String gitToolName(Run<?, ?> run, TaskListener listener) throws IOException, InterruptedException {
+
+        Executor buildExecutor = run.getExecutor();
+        ToolDescriptor<? extends GitTool> toolType;
+        if (buildExecutor != null) {
+            Node currentNode = buildExecutor.getOwner().getNode();
+            //Check node is not null
+            if (currentNode != null) {
```

Comment:
@arpoch always try to break up complicated business logic, it has two clear benefits:
- clean, cognitively simpler code to maintain
- nested branches of `if else` tend to miss edge cases in my experience

for example, if (currentNode != null && currentNode.getNodeName().equals("")) can be combined together, right?

Reply:
Thanks for the tip Rishabh,  but I made some improvements in the code which makes this suggested change outdated.

Reply:
I think I missed one thing, are we making sure that we are only looking for a cli git installation and not a jgit or jgitapache one?